### DISCUSSION TITLE
Markdown Export in Alphabetical order

### DISCUSSION THREAD

**Discussion by SkipToTheEndpoint** [Micke-K/IntuneManagement#90] [General]
The Markdown export feature is obviously experimental, but is there at least an easy way of getting it to work through and document policies in alphabetical order for the produced markdown file?

Minor I know :)

**Comment by Micke-K** [comment]
Hello!

Yes, I have to revisit this. I've been thinking about it how to export it the best way but not taking the time to do anything about it. 

I have to split it because the document gets too large in big environments. I couldn't even open it in any viewer. 

I'll see what can do with sorting them. That sounds like an easy task. I managed to do it for the word output 😄

**Reply by SkipToTheEndpoint** [reply]
Thanks

I noticed #35 after I posted this. Happy to help with testing :)

The one thing I've found tricky with all setting exports is being able to denote a "sub-setting", such as:
> Configure SMB v1 client driver: Enabled
> Configure MrxSmb10 driver: Disable driver (recommended)

They appear as separate settings even though the latter is dependant on the former, and had a question about "Why is the SMB v1 driver being enabled!?" but I'm not sure how would be best to identify or display them.

**Comment by Micke-K** [comment]
It might be possible to fix but not in a quick and easy way. I know there are parent/child relationship between many settings and I think I collect that as well. The problem would be to output it in a good way. And I do collect descriptions for most of the properties as well. 

I'll see if I can get time to look at this. There are a few things in the backlog already. 

Cheers!

**Comment by Micke-K** [comment]
Hello!

Long time. This is just to let you know that this is not forgotten. I am looking into this but it's a lot and the more I dig, the more I find to add/fix. I had to redesign some parts of the documentation feature...

I hope I can have something for you to test in the next few weeks or so. I've been very busy and this is taking longer than I thought but I think you will like it :)

Cheers!

**Comment by Micke-K** [comment]
Hello,

Here is an updated version for you to test. It has LOTS of updates for the documentation feature:

- Rewrote documentation engine for the Settings Catalog - Should be faster and fixes a few issues
- Added HTML Output Provider
- Added output options for MD and HTML (Document name and CSS file)
- Sorted the documented policies
- Added indent for sub-settings. Should be supported for all policy types but it's not 100% tested so let me know if it looks weird
- Added table values (MD and HTML) for multi column values (example the return codes property for Win32 Apps)
- Moved some settings to the Output tab
- Made some settings generic (Document scripts and remove script signature)
- Fixed some documentation issues for Autopilot and Apps
- Added Requirement and Detection script for Win32 Apps
- Support for filtering based on Name or ScopTag for Bulk Document Types

+probably a few other minor things I forgot about...

It is FAR from 100% tested. It's just too many policy types to test it all... The indent functionality looks good but I really expect it to be lots of issues.
Settings Catalog documentation sorts the properties. Not sure in what order MS does it. Looks like they sometimes sort and sometimes don't. Some Settings Catalog policies will look weird with the indent feature but they actually look the same in the portal. Some settings are inside a group but the group itself is not added so the properties will be 1 level in.

Let me know how it goes or if you have any questions

Cheers!

[IntuneManagement.zip](https://github.com/Micke-K/IntuneManagement/files/10121521/IntuneManagement.zip)

**Comment by Micke-K** [comment]
Not sure of you gave tested this yet I found that I have to redesign it a bit more. The curet one adds values based on property type which works until you have a property with multiple values eg Firewall Rules. 

I'll see when I can fix this...

Cheers!

**Reply by SkipToTheEndpoint** [reply]
Sorry! I did manage to test this briefly and the HTML option was great for everything I tried to export. My baseline is almost exclusively Settings Catalog-based.
I'll try and do some better testing with your next iteration.

**Comment by Micke-K** [comment]
A new version to test. I had to rewrite even more parts of the documentation. Some additional features added for MD and HTML. They now support documenting to one file per policy. I also added support for setting indent on sub-settings for Word.
All providers should work so please report any issues. There are still lots of small things to do.
    
[IntuneManagement.zip](https://github.com/Micke-K/IntuneManagement/files/10267169/IntuneManagement.zip)
